### PR TITLE
feat(tree): Control `--charset` via auto-detecting config value

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

This tries to generalize `cargo tree --charset` so any part of cargo's output can use it.  For example,
- `cargo search` could use this for fancier tables
- `cargo add` could use this for fancier feature lists
- #12235 could use this for fancy rendering like miette does (CC @Muscraft )
- Progress bars could use fancier characters <-- this is what I'm personally working towards

This builds on the idea from #12889 that we can use fancier terminal features so long as we have good auto-detection and provide users an escape hatch until the auto-detection is improved or in case they disagree.

As a side benefit
- `cargo tree` will auto-detect when the prior default of `--charset utf8` is a good choice
- Users can control `cargo tree --charset` via `.cargo/config.toml`

### How should we test and review this PR?

This is gradually introduced through the individual commits.

### CODE REVIEW

File: src/doc/src/reference/config.md
Comment (epage) [inline src/doc/src/reference/config.md:1291]: My main concern is the future possibilities.  If we later add [emoji](https://crates.io/crates/terminal-supports-emoji) or [devicon fonts](https://devicon.dev/), how would we represent this?  A separate field, hierarchical values, or a list?

Comment (weihanglo) [reply src/doc/src/reference/config.md:1291]: One possible way: default to "auto" (I sympathize this word after cargo meeting today), and remove the `utf8` option. This could maximize our future compatibility.

**Review by weihanglo — COMMENTED** [review]
Implementation-wise looks good to me 👍🏾.

File: src/doc/src/reference/config.md
Comment (epage) [reply src/doc/src/reference/config.md:1291]: The downside is that doesn't allow forcing of `utf8` if  our `auto` is being overly selective.  Granted, maybe thats not too bad for now?  It would mean we'd need separate enums for config from internal / CLI to support `cargo tree --charset utf8`.

Thoughts?

Comment (weihanglo) [reply src/doc/src/reference/config.md:1291]: Do we have any incoming feature going to use Unicode tables? If not, I feel like stabilizing fewer things buys us more time to deal with future changes.

BTW, I was about to convince myself an “auto” option is more aligned to other `term` configs, until I saw `term.hyperlinks`.

>  It would mean we'd need separate enums for config from internal / CLI to support cargo tree --charset utf8.

Yeah. Sounds like it 😞.

Comment (epage) [reply src/doc/src/reference/config.md:1291]: Yes, there isn't something using this yet.  Overall, this is part of a slow roll out of features to unblock user control over more complex progress bars so we have cleaner command output.

I also don't see us evaluating devicons any time soon.  Emoji is hit or miss. I don't see an immediate use of it but personally not a big emoji person.


Maybe we can at least hold off on this until I get a little further into the progress bar implementation research?  That way if emoji somehow looks useful in that time we re-evaluate?

Comment (epage) [reply src/doc/src/reference/config.md]: I just re-worked this PR to instead have a `term.unicode` field.  This would allow us to add more fields in the future like `term.emoji` without worrying about the specifics of what we might need in the future.

Thoughts?

File: src/cargo/core/shell.rs
Comment (weihanglo) [inline src/cargo/core/shell.rs:558]: nit: While I know if-else is more readable, should we collapse it?

```suggestion
    !stream.is_terminal() || supports_unicode::supports_unicode()
```

**Review by weihanglo — APPROVED** [review]
Looks good to me. `r=weihanglo` whenever you feel ready to merge.

**Comment by epage** [comment]
@rfcbot fcp merge

This adds a new config field, `term.unicode = <bool>`.  When unset, it is auto-detected via [supports-unicode](https://crates.io/crates/supports-unicode).  As of this PR, this just controls the default for  `cargo tree --charset` but this allows us to leverage this anywhere.  Potential ideas include
- `cargo info` feature trees
- Diagnostics like [miette](https://crates.io/crates/miette)
- Fancier looking progress updates

**Comment by weihanglo** [comment]
I don't think this needs a full 10-day FCP. Merge this now.

@bors r+

**Pull request merged**
